library/init/category/lawful.lean

```diff
@@ -0,0 +1,223 @@
+/-
+Copyright (c) 2017 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sebastian Ullrich
+-/
+prelude
+import init.category.monad init.meta.interactive
+import init.category.state init.category.except init.category.reader init.category.cont
+universes u v
+
+open function
+open tactic
+
+meta def control_laws_tac := whnf_target >> intros >> to_expr ``(rfl) >>= exact
+
+class is_lawful_functor (f : Type u → Type v) [functor f] : Prop :=
+(map_const_eq : ∀ {α β : Type u}, @has_map.map_const f _ α β = (<$>) ∘ const β . control_laws_tac)
+-- `functor` is indeed a categorical functor
+(id_map       : Π {α : Type u} (x : f α), id <$> x = x)
+(comp_map     : Π {α β γ : Type u} (g : α → β) (h : β → γ) (x : f α), (h ∘ g) <$> x = h <$> g <$> x)
+
+export is_lawful_functor (map_const_eq id_map comp_map)
+attribute [simp] id_map
+-- `comp_map` does not make a good simp lemma
+
+class is_lawful_applicative (f : Type u → Type v) [applicative f] extends is_lawful_functor f : Prop :=
+(seq_left_eq  : ∀ {α β : Type u} (a : f α) (b : f β), a <* b = const β <$> a <*> b . control_laws_tac)
+(seq_right_eq : ∀ {α β : Type u} (a : f α) (b : f β), a *> b = const α id <$> a <*> b . control_laws_tac)
+-- applicative laws
+(pure_seq_eq_map : ∀ {α β : Type u} (g : α → β) (x : f α), pure g <*> x = g <$> x) -- . control_laws_tac)
```

Comment:
Is the comment `--  . control_laws_tac)` a leftover or a typo?

Reply:
Looks like a leftover from the original definition, removed.